src/mds/MDCache.cc

```diff
@@ -10236,29 +10244,46 @@ void MDCache::scan_stray_dir(dirfrag_t next)
 	continue;
 
       if (!dir->can_auth_pin()) {
-	dir->add_waiter(CDir::WAIT_UNFREEZE, new C_MDC_RetryScanStray(this, dir->dirfrag()));
-	return;
+	dir->add_waiter(CDir::WAIT_UNFREEZE, new C_MDC_RetryScanStray(this, dir->dirfrag(), std::move(cmd_ctx)));
+	return -EAGAIN;
       }
 
       if (!dir->is_complete()) {
-	dir->fetch(new C_MDC_RetryScanStray(this, dir->dirfrag()));
-	return;
+	dir->fetch(new C_MDC_RetryScanStray(this, dir->dirfrag(), std::move(cmd_ctx)));
+	return -EAGAIN;
       }
 
+  
       for (auto &p : dir->items) {
 	CDentry *dn = p.second;
 	dn->state_set(CDentry::STATE_STRAY);
 	CDentry::linkage_t *dnl = dn->get_projected_linkage();
 	if (dnl->is_primary()) {
 	  CInode *in = dnl->get_inode();
+    // only if we came from asok cmd handler
+    if (cmd_ctx) {
+      cmd_ctx->get_formatter()->open_object_section("stray_inode");
+      cmd_ctx->get_formatter()->dump_int("ino: ", in->ino());
+      cmd_ctx->get_formatter()->dump_string("stray_prior_path: ", in->get_inode()->stray_prior_path);
+      in->dump(cmd_ctx->get_formatter(), CInode::DUMP_CAPS);
+      cmd_ctx->get_formatter()->close_section();
+    }
 	  if (in->get_inode()->nlink == 0)
 	    in->state_set(CInode::STATE_ORPHAN);
-	  maybe_eval_stray(in);
+    // no need to evaluate stray when dumping the dir content
+    if (!cmd_ctx) {
+	    maybe_eval_stray(in);
+    }
 	}
       }
     }
     next.frag = frag_t();
   }
+  // only if we came from asok cmd handler
+  if (cmd_ctx) {
+    cmd_ctx->finish(0);
```

Comment:
This is fine but unconventional in the code base. Note: this bypasses `MDSContext::complete` which is not great.

If you want to fortify the safety of the code, I recommend:

Move

https://github.com/ceph/ceph/blob/94254176eac993ec3e3af34bc96cdc253988f110/src/mds/MDSContext.cc#L25-L29

to a new `MDSContext::finish` method. Remove the `MDSContext::complete` overload.

Then change `C_MDC_RetryScanStray::finish` to first call its parent `MDSContext::finish` so we get the checks/prints that we expect.

Reply:
Also please note: this type of refactor I'm recommending should be in a separate commit with an appropriate commit message explaining the change.